x/superfluid/keeper/stake.go

```diff
@@ -269,23 +269,96 @@ func (k Keeper) SuperfluidUndelegate(ctx sdk.Context, sender string, lockID uint
 // SuperfluidUnbondLock unbonds the lock that has been used for superfluid staking.
 // This method would return an error if the underlying lock is not superfluid undelegating.
 func (k Keeper) SuperfluidUnbondLock(ctx sdk.Context, underlyingLockId uint64, sender string) error {
+	_, err := k.unbondLock(ctx, underlyingLockId, sender, sdk.Coins{})
+	return err
+}
+
+// SuperfluidUndelegateAndUnbondLock unbonds partial or full amount from the
+// underlying lock that has been used for superfluid staking.
+// This method returns the lock id, same lock id if unlock amount is equal to the
+// underlying lock amount. Otherwise it returns the newly created lock id.
+func (k Keeper) SuperfluidUndelegateAndUnbondLock(ctx sdk.Context, lockID uint64, sender string, amount sdk.Int) (uint64, error) {
+	lock, err := k.lk.GetLockByID(ctx, lockID)
+	if err != nil {
+		return 0, err
+	}
+
+	coins := sdk.Coins{sdk.NewCoin(lock.Coins[0].Denom, amount)}
+	if coins[0].IsZero() {
+		return 0, fmt.Errorf("amount to unlock must be greater than 0")
+	}
+	if lock.Coins[0].IsLT(coins[0]) {
+		return 0, fmt.Errorf("requested amount to unlock exceeds locked tokens")
+	}
+
+	// get intermediary account before connection is deleted in SuperfluidUndelegate
+	intermediaryAcc, found := k.GetIntermediaryAccountFromLockId(ctx, lockID)
+	if !found {
+		return 0, types.ErrNotSuperfluidUsedLockup
+	}
+
+	// undelegate all
+	err = k.SuperfluidUndelegate(ctx, sender, lockID)
+	if err != nil {
+		return 0, err
+	}
+
+	// unbond partial or full locked amount
+	newLockID, err := k.unbondLock(ctx, lockID, sender, coins)
+	if err != nil {
+		return 0, err
+	}
+
+	// check new lock id, return if unbond amount = locked amount
+	if lock.Coins[0].IsEqual(coins[0]) {
+		if newLockID != lockID {
+			panic(fmt.Errorf("expected new lock id %v to = lock id %v", newLockID, lockID))
+		}
+		return lock.ID, nil
+	} else {
+		if newLockID == lockID {
+			panic(fmt.Errorf("expected new lock id %v to != lock id %v", newLockID, lockID))
+		}
+	}
+
+	// delete synthetic unlocking lock created in the last step of SuperfluidUndelegate
+	synthdenom := unstakingSyntheticDenom(lock.Coins[0].Denom, intermediaryAcc.ValAddr)
+	err = k.lk.DeleteSyntheticLockup(ctx, lockID, synthdenom)
+	if err != nil {
+		return 0, err
+	}
+
+	// re-delegate remainder
+	err = k.SuperfluidDelegate(ctx, sender, lockID, intermediaryAcc.ValAddr)
```

Comment:
I think lockID is correct, but there might be something I misunderstood about the code.

The underlying lock will be split only if the unlock amount is less than the locked amount.
The newly split lock will be an unlocking lock. Original lock will still be locked.

So inside superfluid we need to update in accordance with what happened in the lockup module.
If `lockID = newLockID`, then no lock was split, so the code exits without re-delegating

https://github.com/stakewithus/osmosis/blob/9fcd3c123967f277c90c7ad2f66809568092b5af/x/superfluid/keeper/stake.go#L317

Otherwise `lockID != newLockID`, so a new unlocking lock was created.
`lockID` contains the amount still locked in the `lockup` module.
`newLockID` contains the amount unlocked.

So we re-delegate `lockID`, updating the amount to superfluid stake, and create a synthetic unlocking lock for `newLockID`

Reply:
I see I see, this makes sense! Would it be possible to add some of these explanations as inline comments?